### PR TITLE
Handle constraints being applied to schemas that don't accept it

### PR DESCRIPTION
**Pull request by adriangb**
This handles the particular case of `pattern` not working when using a custom `GenerateSchema` (reported in https://github.com/pydantic/pydantic/issues/6045#issuecomment-1652440886) but also tries to be more general to handle other similar issues.

Selected Reviewer: @samuelcolvin

### CODE REVIEW

**Comment by adriangb** [comment]
Open questions:
- Should we try to error fast if we _know_ there's going to be an issue, e.g. `Annotated[int, Field(pattern=...)]`. The only way I see to do this is essentially special case by checking `constraint = 'pattern' and schema['type'] in ('int', 'float', etc.)`
- Can we get rid of the "merging" of `FieldInfo`? I think for validation yes, for JSON schema we may still need to merge them, but not even sure about that.
- Do we still want to make a more performant version of this in pydantic-core?

**Comment by adriangb** [comment]
please review

**Comment by adriangb** [comment]
This is now working and allowing constraint re-ordering. The cost is running things in Python when they can't be "inlined" into Rust. There's also a slight inconsistency I belive since when we _do_ "inline" it the order is not preserved (it's whatever the order of the if statements is in Rust). I guess that's fine.

File: tests/test_types.py [hunk 1]
```diff
@@ -1675,25 +1675,6 @@ class Model(BaseModel):
...
-def test_invalid_schema_constraints(kwargs, type_):
-    match = (
-        r'(:?Invalid Schema:\n.*\n  Extra inputs are not permitted)|(:?The following constraints cannot be applied to)'
-    )
-    with pytest.raises((SchemaError, TypeError), match=match):
-
-        class Foo(BaseModel):
-            a: type_ = Field('foo', title='A title', description='A description', **kwargs)
-
-
```
Comment (dmontagu) [inline tests/test_types.py:1696]:
```suggestion
@pytest.mark.parametrize(
    'kwargs,type_',
    [
        ({'gt': 0}, conlist(int, min_length=4)),
        ({'gt': 0}, conset(int, min_length=4)),
        ({'gt': 0}, confrozenset(int, min_length=4)),
    ],
)
def test_invalid_schema_constraints(kwargs, type_):
    match = (
        r'(:?Invalid Schema:\n.*\n  Extra inputs are not permitted)|(:?The following constraints cannot be applied to)'
    )
    with pytest.raises((SchemaError, TypeError), match=match):

        class Foo(BaseModel):
            a: type_ = Field('foo', title='A title', description='A description', **kwargs)
```

**Review by dmontagu — APPROVED** [review]

File: tests/test_types.py (hunk 1 above)
Comment (adriangb) [reply tests/test_types.py:1696]: The reason why the pattern one doesn't give an error anymore and the others do is that those go down the `__prepare_pydantic_annotations__` path. One of the important bits of this PR is that we might be able to get rid of that in the future, but I guess it makes sense to leave the tests in for now.

**Comment by adriangb** [comment]
These two give different results:

```python
from typing_extensions import Annotated

from pydantic import BaseModel, Field, AfterValidator

from pydantic.type_adapter import TypeAdapter


Tp = Annotated[int, AfterValidator(lambda x: x * 2), Field(gt=3)]

class A(BaseModel):
    x: Tp

print(A.__pydantic_core_schema__['definitions'][0]['schema']['fields']['x']['schema'])

ta = TypeAdapter(Tp)
print(ta.core_schema)
```

It seems like this is because BaseModel merges field infos when it's constructed. We'd have to get rid of all field-info merging to get the right behavior w.r.t. ordering of constraints. I'm inclined to leave fixing that for a future PR. WDYT @dmontagu?

File: pydantic/_internal/_known_annotated_metadata.py [hunk 2]
```diff
@@ -28,15 +29,57 @@
...
+for constraint in TIMEDELTA_CONSTRAINTS:
+    CONSTRAINTS_TO_ALLOWED_SCHEMAS[constraint].update(('timedelta',))
+for constraint in TIME_CONSTRAINTS:
+    CONSTRAINTS_TO_ALLOWED_SCHEMAS[constraint].update(('time',))
+for schema_type in (*TEXT_SCHEMA_TYPES, *SEQUENCE_SCHEMA_TYPES, *NUMERIC_SCHEMA_TYPES, 'typed-dict', 'model'):
+    CONSTRAINTS_TO_ALLOWED_SCHEMAS['strict'].add(schema_type)
+for constraint in URL_CONSTRAINTS:
+    CONSTRAINTS_TO_ALLOWED_SCHEMAS[constraint].update(('url', 'multi-host-url'))
+for constraint in BOOL_CONSTRAINTS:
+    CONSTRAINTS_TO_ALLOWED_SCHEMAS[constraint].update(('bool',))
```
Comment (samuelcolvin) [inline pydantic/_internal/_known_annotated_metadata.py:81]: does this get called on `from pydantic import X`? If so we might need to worry about import time, if it's later/when require it shouldn't be a problem.

File: pydantic/_internal/_known_annotated_metadata.py [hunk 3]
```diff
@@ -96,82 +139,150 @@ def apply_known_metadata(annotation: Any, schema: CoreSchema) -> CoreSchema | No
...
+                schema,
+            )
+        elif constraint == 'pattern':
+            # insert a str schema to make sure the regex engine matches
+            return cs.chain_schema(
+                [
+                    schema,
+                    cs.str_schema(pattern=value),
+                ]
+            )
```
Comment (samuelcolvin) [inline pydantic/_internal/_known_annotated_metadata.py:165]: these can probably be made more compact thereby making the function easier to read.

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM.

**Comment by samuelcolvin** [comment]
please update.

File: pydantic/_internal/_known_annotated_metadata.py (hunk 3 above)
Comment (adriangb) [reply pydantic/_internal/_known_annotated_metadata.py:165]: black controls formatting not me 😄

File: pydantic/_internal/_known_annotated_metadata.py (hunk 2 above)
Comment (adriangb) [reply pydantic/_internal/_known_annotated_metadata.py:81]: I believe it is imported immediately. It's not all that complex though, just assigning times to a dict. I could hide it behind an lru_cache function or something, but that would probably make schema generation slower (one more dict lookup), and that seems to be a bigger problem right now than import time

File: pydantic/_internal/_validators.py [hunk 4]
```diff
@@ -269,3 +270,12 @@ def max_length_validator(x: Any, max_length: Any) -> Any:
...
         )
     return x
+
+
+def forbid_inf_nan_check(x: Any) -> Any:
+    if not math.isfinite(x):
+        raise PydanticKnownError(
+            'finite_number',
+            {},
+        )
```
[the remark this answers is not in the document]
Comment (adriangb) [reply pydantic/_internal/_validators.py]: yep thanks

**Comment by adriangb** [comment]
please review

File: tests/test_types.py (hunk 1 above)
Comment (adriangb) [reply tests/test_types.py:1696]: I'll add this but just to be clear it's more documenting current behavior than testing the desired / future behavior

**Comment by adriangb** [comment]
please review

File: tests/test_serialize.py [hunk 5]
```diff
@@ -896,7 +896,7 @@ class UserLogin(User):
         password: SecretStr
 
     class OuterModel(BaseModel):
-        maybe_as_any: Optional[SerializeAsAny[User]]
+        maybe_as_any: Optional[SerializeAsAny[User]] = None
```
Comment (adriangb) [inline tests/test_serialize.py:899]: Not strictly necessary for this PR but the insert_assert() below doesn't work without it

File: pydantic/_internal/_validators.py [hunk 6]
```diff
@@ -269,3 +270,9 @@ def max_length_validator(x: Any, max_length: Any) -> Any:
             {'field_type': 'Value', 'max_length': max_length, 'actual_length': len(x)},
         )
     return x
+
+
+def forbid_inf_nan_check(x: Any) -> Any:
+    if not math.isfinite(x):
+        raise PydanticKnownError('finite_number')
+    return x
```
Comment (adriangb) [inline pydantic/_internal/_validators.py:278]: Added a test for the error case here

File: pydantic/_internal/_known_annotated_metadata.py [hunk 7]
```diff
@@ -96,82 +139,150 @@ def apply_known_metadata(annotation: Any, schema: CoreSchema) -> CoreSchema | No
...
+            predicate_name = f'{annotation.func.__qualname__} ' if hasattr(annotation.func, '__qualname__') else ''
+
+            def val_func(v: Any) -> Any:
+                # annotation.func may also raise an exception, let it pass through
+                if not annotation.func(v):
+                    raise PydanticCustomError(
+                        'predicate_failed',
+                        f'Predicate {predicate_name} failed',  # type: ignore
+                        {},
+                    )
```
Comment (adriangb) [inline pydantic/_internal/_known_annotated_metadata.py:281]:
```suggestion
                    raise PydanticCustomError(
                        'predicate_failed',
                        f'Predicate {predicate_name} failed',  # type: ignore
                    )
```

**Review by dmontagu — APPROVED** [review]

**Pull request merged**